client/src/sync.rs

```diff
@@ -0,0 +1,831 @@
+}
+
+/// Policy for selecting between two instances (e.g. client states or vaults) with the same id.
+#[derive(Debug, Clone, Copy)]
+pub enum SelectOrMerge<T> {
+    KeepOld,
+    Replace,
+    /// Recursively merge inner entries.
+    Merge(T),
+}
+
+/// Function for mapping the hierarchy of one [`MergeLayer`] instance to another.
+/// In case of a partial sync, this function can return [`None`] so that this entry is skipped.
+#[derive(Debug, Clone)]
+pub struct Mapper<T> {
+    f: fn(T) -> Option<T>,
+}
+
+impl<T> Default for Mapper<T> {
+    fn default() -> Self {
+        Self { f: |t| Some(t) }
+    }
+}
+
+impl<T> From<fn(T) -> Option<T>> for Mapper<T> {
+    fn from(f: fn(T) -> Option<T>) -> Self {
+        Self { f }
+    }
+}
+
+impl<T> Mapper<T> {
+    pub fn new(&self, f: fn(T) -> Option<T>) -> Self {
+        Self { f }
+    }
+
+    pub fn map(&self, t: T) -> Option<T> {
+        let f = self.f;
+        f(t)
+    }
+}
+
+pub type MergeClientsMapper = Mapper<(VaultId, RecordId)>;
+pub type MergeSnapshotsMapper = Mapper<(ClientId, VaultId, RecordId)>;
+
+pub struct SyncClientState<'a> {
+    pub db: &'a mut DbView<Provider>,
+    pub keystore: &'a mut HashMap<VaultId, Key<Provider>>,
+}
+
+impl<'a> From<&'a mut SecureClient> for SyncClientState<'a> {
+    fn from(client: &'a mut SecureClient) -> Self {
+        SyncClientState {
+            db: &mut client.db,
+            keystore: &mut client.keystore.store,
+        }
+    }
+}
+
+impl<'a, T> From<&'a mut (HashMap<VaultId, Key<Provider>>, DbView<Provider>, T)> for SyncClientState<'a> {
+    fn from((keystore, db, _): &'a mut (HashMap<VaultId, Key<Provider>>, DbView<Provider>, T)) -> Self {
+        SyncClientState { db, keystore }
+    }
+}
+
+/// Merge two client states.
+impl<'a> MergeLayer<'a> for SyncClientState<'a> {
+    type Hierarchy = HashMap<VaultId, Vec<(RecordId, BlobId)>>;
+    type Exported = HashMap<VaultId, Vec<(RecordId, Record)>>;
+    type Path = (VaultId, RecordId);
+    type MergePolicy = SelectOrMerge<SelectOne>;
+    type KeyProvider = HashMap<VaultId, Key<Provider>>;
+
+    fn get_hierarchy(&mut self) -> Result<Self::Hierarchy, VaultError> {
+        let mut map = HashMap::new();
+        for vid in self.db.list_vaults() {
+            let key = self.keystore.get(&vid).ok_or(VaultError::VaultNotFound(vid))?;
+            match self.db.list_records_with_blob_id(key, vid) {
+                Ok(list) => {
+                    map.insert(vid, list);
+                }
+                Err(VaultError::VaultNotFound(_)) => {}
+                e => panic!("{:?}", e),
+            }
+        }
+        Ok(map)
+    }
+
+    fn get_diff(
+        &mut self,
+        other: Self::Hierarchy,
+        mapper: Option<&Mapper<Self::Path>>,
+        merge_policy: &Self::MergePolicy,
+    ) -> Result<Self::Hierarchy, VaultError> {
+        let mut diff = HashMap::<VaultId, Vec<_>>::new();
+        for (vid0, records) in other {
+            for (rid0, bid) in records {
```

Comment:
I wonder, if `BlobId`s can be left out of the synchronization part. I would assume, that the relevant part is the id of the record, since the contents might be outdated, and should be synchronized, otherwise this could be part of a `MergePolicy`, than source is being retained, target being update etc...

Reply:
Ok. Got it. you later differentiate according to the `SelectOrMerge` named policy. Maybe split it into multiple functions each having distinct functionality